deep_quoridor/src/agents/alphazero/alphazero.py

```diff
@@ -546,56 +563,88 @@ def _log_action(
         self.action_log.action_score_ranking(scores)
         self.action_log.action_text(root_action, f"{root_value:0.2f}")
 
-    def get_action(self, observation) -> int:
-        game, player, _ = construct_game_from_observation(observation["observation"])
-        # Run MCTS to get action visit counts
-        root_children, root_value = self.mcts.search(game)
-        visit_counts = np.array([child.visit_count for child in root_children])
-        visit_counts_sum = np.sum(visit_counts)
-        if visit_counts_sum == 0:
-            raise RuntimeError("No nodes visited during MCTS")
-
-        visit_probs = visit_counts / visit_counts_sum
-        self._log_action(
-            visit_probs, root_children, float(root_value), MoveAction(game.board.get_player_position(player))
-        )
+    def get_action_batch(self, observations_with_ids: list[tuple[int, dict]]) -> list[tuple[int, int]]:
+        """
+        Get actions for multiple observations.  Each entry in observation_with_ids needs a distinct id that can be
+        an arbitrary number. The same id will be returned with the action matching the observation.
+        """
+        if not observations_with_ids:
+            return []
+
+        games = []
+        players = []
+        game_indices = []
+        for game_idx, observation in observations_with_ids:
+            g, p, _ = construct_game_from_observation(observation["observation"])
+            games.append(g)
+            players.append(p)
+            game_indices.append(game_idx)
+
+        root_children_batch, root_value_batch = self.mcts.search_batch(games)
```

Comment:
this is the main change for this method: instead of doing mcts search for just one game, we do it for many games, and we get arrays of results.  Below, we iterate through those results, but hat we end up doing is almost the same.  It's hard to follow what changed in the diff